src/QMCWaveFunctions/Fermion/DiracDeterminant.cpp

```diff
@@ -166,30 +164,32 @@ typename DiracDeterminant<DU_TYPE>::PsiValueType DiracDeterminant<DU_TYPE>::rati
                                                                                               GradType& grad_iat,
                                                                                               ComplexType& spingrad_iat)
 {
-  SPOVGLTimer.start();
-  Phi->evaluateVGL(P, iat, psiV, dpsiV, d2psiV);
-  Phi->evaluate_spin(P, iat, psiV, dspin_psiV);
-  SPOVGLTimer.stop();
-
-  UpdateMode = ORB_PBYP_PARTIAL;
-  RatioTimer.start();
-  const int WorkingIndex = iat - FirstIndex;
-  assert(WorkingIndex >= 0);
-  // This is an optimization.
-  // check invRow_id against WorkingIndex to see if getInvRow() has been called already
-  // Some code paths call evalGrad before calling ratioGrad.
-  if (invRow_id != WorkingIndex)
   {
-    invRow_id = WorkingIndex;
-    updateEng.getInvRow(psiM, WorkingIndex, invRow);
+    ScopedTimer local_timer(&SPOVGLTimer);
+    Phi->evaluateVGL(P, iat, psiV, dpsiV, d2psiV);
+    Phi->evaluate_spin(P, iat, psiV, dspin_psiV);
   }
-  curRatio = simd::dot(invRow.data(), psiV.data(), invRow.size());
-  grad_iat += static_cast<ValueType>(static_cast<PsiValueType>(1.0) / curRatio) *
-      simd::dot(invRow.data(), dpsiV.data(), invRow.size());
 
-  spingrad_iat += static_cast<ValueType>(static_cast<PsiValueType>(1.0) / curRatio) *
-      simd::dot(invRow.data(), dspin_psiV.data(), invRow.size());
-  RatioTimer.stop();
+  {
+    ScopedTimer local_timer(&RatioTimer);
+    UpdateMode             = ORB_PBYP_PARTIAL;
+    const int WorkingIndex = iat - FirstIndex;
+    assert(WorkingIndex >= 0);
+    // This is an optimization.
+    // check invRow_id against WorkingIndex to see if getInvRow() has been called already
+    // Some code paths call evalGrad before calling ratioGrad.
```

Comment:
Are there code paths that we can get rid of / simplify in that case? What is the use case for both calls?

Reply:
Calling evalGrad before ratioGrad is required by VMC driver and DMC. To remove this if-statement, we need to inject evalGrad where there is ratioGrad. So this is actually more of a safety mechanism.

Reply:
I will update the notes in an upcoming PR.